app/src/main/java/org/thoughtcrime/securesms/gcm/FcmUtil.java

```diff
@@ -0,0 +1,35 @@
+package org.thoughtcrime.securesms.gcm;
+
+import android.text.TextUtils;
+
+import androidx.annotation.WorkerThread;
+
+import com.google.android.gms.tasks.Tasks;
+import com.google.firebase.messaging.FirebaseMessaging;
+
+import org.signal.core.util.logging.Log;
+import org.whispersystems.libsignal.util.guava.Optional;
+
+import java.util.concurrent.ExecutionException;
+
+public final class FcmUtil {
+
+  private static final String TAG = Log.tag(FcmUtil.class);
+
+  /**
+   * Retrieves the current FCM token. If one isn't available, it'll be generated.
+   */
+  @WorkerThread
+  public static Optional<String> getToken() {
+    String token = null;
+    try {
+      token = Tasks.await(FirebaseMessaging.getInstance().getToken());
+    } catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)